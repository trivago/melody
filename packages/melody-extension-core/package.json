{
  "name": "melody-extension-core",
  "version": "1.0.4",
  "description": "",
  "main": "./lib/index.js",
  "jsnext:main": "./src/index.js",
  "scripts": {
    "build": "mkdir lib; SUPPORT_CJS=true rollup -c ../../rollup.config.js -i src/index.js -o lib/index.js"
  },
  "author": "",
  "license": "Apache-2.0",
  "dependencies": {
    "babel-template": "^6.8.0",
    "babel-types": "^6.8.1",
    "lodash": "^4.12.0",
    "shortid": "^2.2.6"
  },
  "bundledDependencies": [
    "babel-types",
    "babel-template",
    "shortid"
  ],
  "peerDependencies": {
    "melody-idom": "^1.0.0",
    "melody-parser": "^1.0.0",
    "melody-runtime": "^1.0.0",
    "melody-traverse": "^1.0.0",
    "melody-types": "^1.0.0"
  },
  "devDependencies": {
    "rollup-plugin-babel": "^2.6.1"
  }
}
