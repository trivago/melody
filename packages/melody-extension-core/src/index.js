/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { unaryOperators, binaryOperators, tests } from './operators';
import { AutoescapeParser } from './parser/autoescape';
import { BlockParser } from './parser/block';
import { DoParser } from './parser/do';
import { EmbedParser } from './parser/embed';
import { ExtendsParser } from './parser/extends';
import { FilterParser } from './parser/filter';
import { FlushParser } from './parser/flush';
import { ForParser } from './parser/for';
import { FromParser } from './parser/from';
import { IfParser } from './parser/if';
import { ImportParser } from './parser/import';
import { IncludeParser } from './parser/include';
import { MacroParser } from './parser/macro';
import { SetParser } from './parser/set';
import { SpacelessParser } from './parser/spaceless';
import { UseParser } from './parser/use';
import { MountParser } from './parser/mount';

import forVisitor from './visitors/for';
import testVisitor from './visitors/tests';
import filters from './visitors/filters';
import functions from './visitors/functions';

const filterMap = [
    'attrs',
    'classes',
    'styles',
    'batch',
    'escape',
    'format',
    'merge',
    'nl2br',
    'number_format',
    'raw',
    'replace',
    'reverse',
    'round',
    'striptags',
    'title',
    'url_encode',
    'trim',
].reduce((map, filterName) => {
    map[filterName] = 'melody-runtime';
    return map;
}, Object.create(null));

Object.assign(filterMap, filters);

const functionMap = [
    'attribute',
    'constant',
    'cycle',
    'date',
    'max',
    'min',
    'random',
    'range',
    'source',
    'template_from_string',
].reduce((map, functionName) => {
    map[functionName] = 'melody-runtime';
    return map;
}, Object.create(null));
Object.assign(functionMap, functions);

export const extension = {
    tags: [
        AutoescapeParser,
        BlockParser,
        DoParser,
        EmbedParser,
        ExtendsParser,
        FilterParser,
        FlushParser,
        ForParser,
        FromParser,
        IfParser,
        ImportParser,
        IncludeParser,
        MacroParser,
        SetParser,
        SpacelessParser,
        UseParser,
        MountParser,
    ],
    unaryOperators,
    binaryOperators,
    tests,
    visitors: [forVisitor, testVisitor],
    filterMap,
    functionMap,
};

export {
    AutoescapeBlock,
    BlockStatement,
    BlockCallExpression,
    MountStatement,
    DoStatement,
    EmbedStatement,
    ExtendsStatement,
    FilterBlockStatement,
    FlushStatement,
    ForStatement,
    ImportDeclaration,
    FromStatement,
    IfStatement,
    IncludeStatement,
    MacroDeclarationStatement,
    VariableDeclarationStatement,
    SetStatement,
    SpacelessBlock,
    AliasExpression,
    UseStatement,
    UnaryNotExpression,
    UnaryNeqExpression,
    UnaryPosExpression,
    BinaryOrExpression,
    BinaryAndExpression,
    BitwiseOrExpression,
    BitwiseXorExpression,
    BitwiseAndExpression,
    BinaryEqualsExpression,
    BinaryNotEqualsExpression,
    BinaryLessThanExpression,
    BinaryGreaterThanExpression,
    BinaryLessThanOrEqualExpression,
    BinaryGreaterThanOrEqualExpression,
    BinaryNotInExpression,
    BinaryInExpression,
    BinaryMatchesExpression,
    BinaryStartsWithExpression,
    BinaryEndsWithExpression,
    BinaryRangeExpression,
    BinaryAddExpression,
    BinaryMulExpression,
    BinaryDivExpression,
    BinaryFloorDivExpression,
    BinaryModExpression,
    BinaryPowerExpression,
    BinaryNullCoalesceExpression,
    TestEvenExpression,
    TestOddExpression,
    TestDefinedExpression,
    TestSameAsExpression,
    TestNullExpression,
    TestDivisibleByExpression,
    TestConstantExpression,
    TestEmptyExpression,
    TestIterableExpression,
} from './types';
