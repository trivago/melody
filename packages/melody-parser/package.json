{
  "name": "melody-parser",
  "version": "1.0.4",
  "description": "",
  "main": "./lib/index.js",
  "jsnext:main": "./src/index.js",
  "scripts": {
    "build": "mkdir lib; rollup -c ../../rollup.config.js -i src/index.js -o lib/index.js"
  },
  "author": "",
  "license": "Apache-2.0",
  "dependencies": {
    "he": "^1.1.0",
    "lodash": "^4.12.0",
    "melody-code-frame": "1.0.4"
  },
  "bundledDependencies": [
    "he"
  ],
  "peerDependencies": {
    "melody-types": "^1.0.0"
  },
  "devDependencies": {
    "rollup-plugin-babel": "^2.6.1"
  }
}
