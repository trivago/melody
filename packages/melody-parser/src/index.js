/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import Parser from './Parser';
import TokenStream from './TokenStream';
import * as Types from './TokenTypes';
import Lexer from './Lexer';
import { EOF, CharStream } from './CharStream';
import { LEFT, RIGHT } from './Associativity';
import {
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    isMelodyExtension,
} from './util';

function parse(code, options, ...extensions) {
    return createExtendedParser(code, options, ...extensions).parse();
}

function createExtendedParser(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedParser(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = createExtendedLexer(code, options, ...passedExtensions);
    const parser = new Parser(
        new TokenStream(lexer, passedOptions),
        passedOptions
    );
    for (const ext of passedExtensions) {
        parser.applyExtension(ext);
    }
    return parser;
}

function createExtendedLexer(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedLexer(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = new Lexer(new CharStream(code), passedOptions);
    for (const ext of passedExtensions) {
        lexer.applyExtension(ext);
    }
    return lexer;
}

export {
    Parser,
    TokenStream,
    Lexer,
    EOF,
    CharStream,
    LEFT,
    RIGHT,
    parse,
    createExtendedLexer,
    createExtendedParser,
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    Types,
};
