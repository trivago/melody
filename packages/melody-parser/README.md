# melody-parser

This parser is an extensible parser for the Twig template language. It takes source code as input and produces an abstract syntax tree (AST) as output.

## Usage

```
const { parse } = require('melody-parser');

const code = '{% spaceless %} This is some Twig code {% endspaceless %}';
const abstractSyntaxTree = parse(code);
```

The `parse` function is a convenient fa√ßade function that does the instantiation of Lexer, Tokenizer, and Parser, and passes forward any options that may be provided. Example with options:

```javascript
const abstractSyntaxTree = parse(code, {
    ignoreComments: false,
    ignoreHtmlComments: false,
    decodeEntities: true,
});
```

There is also a third parameter called `extensions`. It collects the third parameter and all parameters that come after it, so you can pass as many extensions as you want:

```javascript
import { extension as coreExtensions } from 'melody-extension-core';
import customExtension from 'melody-extension-custom';

const abstractSyntaxTree = parse(
    code,
    {
        decodeEntities: true,
    },
    coreExtension,
    customExtension
);
```

If you have no options to pass, you can also omit them, and start passing extensions from the 2nd parameter on:

```javascript
import { extension as coreExtensions } from 'melody-extension-core';
import customExtension from 'melody-extension-custom';

const abstractSyntaxTree = parse(code, coreExtension, customExtension);
```

## Options

### ignoreComments (`true`)

If set to `true`, Twig comments will not be part of the resulting abstract syntax tree (AST). Defaults to `true`.

### ignoreHtmlComments (`true`)

If set to `true`, HTML comments will not be part of the resulting abstract syntax tree (AST). Defaults to `true`.

### preserveSourceLiterally (`false`)

This option can be useful for pretty printing scenarios, hence it is turned off by default. If set to `true`, transformations like decoding of HTML entities or escape sequences will not be applied.

### decodeEntities (`true`, deprecated, use `preserveSourceLiterally` instead)

Character references/entities like `&#8206;` will be decoded if this is set to `true` (default). Otherwise, the string will be taken over verbatim to the AST.

### ignoreDeclarations (`true`)

Declarations like `<!DOCTYPE html>` are not added to the parse tree if set to `true`.
