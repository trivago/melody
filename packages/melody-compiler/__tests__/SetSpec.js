/**
 * Copyright 2019 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { parse, getNodeSource } from 'melody-parser';
import { extension } from 'melody-extension-core';

describe('set', function() {
    test('should parse set...endset', function() {
        const node = parse(
            `{% set foo -%}
            <p>Some more text</p>
        {%- endset %}`,
            { applyWhitespaceTrimming: false },
            extension
        );
        expect(node.expressions[0]).toMatchSnapshot();
    });

    test('should parse a simple set statement', function() {
        const node = parse(
            `{%- set foo = 0 -%}`,
            { applyWhitespaceTrimming: false },
            extension
        );
        expect(node.expressions[0]).toMatchSnapshot();
    });

    test('should be able to reproduce the original source', function() {
        const setSource = '{% set foo = 0 %}';
        const source = `abc${setSource}def`;
        const node = parse(source, extension);
        expect(getNodeSource(node.expressions[1], source)).toEqual(setSource);
    });
});
