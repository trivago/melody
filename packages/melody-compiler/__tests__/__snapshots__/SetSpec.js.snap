// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set should parse a simple set statement 1`] = `
Object {
  "assignments": Array [
    Object {
      "name": Object {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "VariableDeclarationStatement",
      "value": Object {
        "type": "NumericLiteral",
        "value": 0,
      },
    },
  ],
  "tagNameLoc": Object {
    "column": 4,
    "index": 4,
    "line": 1,
  },
  "trimLeft": true,
  "trimLeftEndset": false,
  "trimRight": true,
  "trimRightSet": false,
  "type": "SetStatement",
}
`;

exports[`set should parse set...endset 1`] = `
Object {
  "assignments": Array [
    Object {
      "name": Object {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "VariableDeclarationStatement",
      "value": Array [
        Object {
          "type": "PrintTextStatement",
          "value": Object {
            "type": "StringLiteral",
            "value": "
            ",
          },
        },
        Object {
          "attributes": Array [],
          "children": Array [
            Object {
              "type": "PrintTextStatement",
              "value": Object {
                "type": "StringLiteral",
                "value": "Some more text",
              },
            },
          ],
          "elementNameLoc": Object {
            "column": 13,
            "index": 28,
            "line": 2,
          },
          "name": "p",
          "selfClosing": false,
          "type": "Element",
        },
        Object {
          "type": "PrintTextStatement",
          "value": Object {
            "type": "StringLiteral",
            "value": "
        ",
          },
        },
      ],
    },
  ],
  "tagNameLoc": Object {
    "column": 3,
    "index": 3,
    "line": 1,
  },
  "trimLeft": false,
  "trimLeftEndset": true,
  "trimRight": false,
  "trimRightSet": true,
  "type": "SetStatement",
}
`;
